traffic_ops/traffic_ops_golang/origin/origins.go

```diff
@@ -487,3 +488,277 @@ func deleteQuery() string {
+}
+
+// Update a Origin for APIv5.
+func Update(w http.ResponseWriter, r *http.Request) {
+	inf, sysError, userError, errorCode := api.NewInfo(r, []string{"id"}, []string{"id"})
+	tx := inf.Tx
+	if sysError != nil || userError != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errorCode, userError, sysError)
+		return
+	}
+	defer inf.Close()
+
+	origin, readValErr := readAndValidateJsonStruct(r)
+	if readValErr != nil {
+		api.HandleErr(w, r, tx.Tx, http.StatusBadRequest, readValErr, nil)
+		return
+	}
+	userErr, sysErr, errCode := checkTenancy(&origin.TenantID, &origin.DeliveryServiceID, tx, inf.User)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	isPrimary := false
+	ds := 0
+	var existingLastUpdated time.Time
+
+	requestedOriginId := inf.IntParams["id"]
+
+	q := `SELECT is_primary, deliveryservice, last_updated FROM origin WHERE id = $1`
+	errLookup := tx.QueryRow(q, requestedOriginId).Scan(&isPrimary, &ds, &existingLastUpdated)
+	if errLookup != nil {
+		if errors.Is(errLookup, sql.ErrNoRows) {
+			api.HandleErr(w, r, tx.Tx, http.StatusNotFound, fmt.Errorf("origin not found"), nil)
+			return
+		}
+		api.HandleErr(w, r, tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("database error: %w, when checking if origin with id %d exists", errLookup, requestedOriginId))
+		return
+	}
+	// check if the entity was already updated
+	userErr, sysErr, errCode = api.CheckIfUnModified(r.Header, inf.Tx, requestedOriginId, "origin")
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	if isPrimary && origin.DeliveryServiceID != ds {
+		api.HandleErr(w, r, tx.Tx, http.StatusBadRequest, fmt.Errorf("cannot update the delivery service of a primary origin"), nil)
+		return
+	}
+
+	_, cdnName, _, err := dbhelpers.GetDSNameAndCDNFromID(tx.Tx, origin.DeliveryServiceID)
+	if err != nil {
+		api.HandleErr(w, r, tx.Tx, http.StatusInternalServerError, err, nil)
+		return
+	}
+	userErr, sysErr, errCode = dbhelpers.CheckIfCurrentUserCanModifyCDN(tx.Tx, string(cdnName), inf.User.UserName)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+
+	var lastUpdated time.Time
+	query := `UPDATE origin SET
+					cachegroup=$1,
+					coordinate=$2,
+					deliveryservice=$3,
+					fqdn=$4,
+					ip6_address=$5,
+					ip_address=$6,
+					name=$7,
+					port=$8,
+					profile=$9,
+					protocol=$10,
+					tenant=$11
+					WHERE id=$12 RETURNING last_updated`
+	errUpdate := tx.QueryRow(query, origin.CachegroupID, origin.CoordinateID, origin.DeliveryServiceID,
+		origin.FQDN, origin.IP6Address, origin.IPAddress, origin.Name,
+		origin.Port, origin.ProfileID, origin.Protocol, origin.TenantID, requestedOriginId).Scan(&lastUpdated)
+	if errUpdate != nil {
+		if errors.Is(errUpdate, sql.ErrNoRows) {
+			api.HandleErr(w, r, tx.Tx, http.StatusBadRequest, fmt.Errorf("origin: %d not found", requestedOriginId), nil)
```

Comment:
If something isn't found, we should use `http.StatusNotFound` as the response status code